### PR TITLE
New features

### PR DESCRIPTION
**Pull request by calevans**
Added to Twitter:
- friendsIds
- listsSubscribers

Also, added ```getRateLimit()``` to Response and a protected array to hold the information passed back in via headers.

### CODE REVIEW

File: library/ZendService/Twitter/Twitter.php [hunk 1]
```diff
@@ -1355,4 +1355,65 @@ protected function friendsIds($id, array $params = [])
...
+     * Returns the members of the specified list. Private list members will 
+     * only be shown if the authenticated user owns the specified list.
+     *
+     * Returns the next cursor if there are more to be returned.
+     *
+     * @param  int|string $id
+     * @param  array $params
+     * @return Response
+     */
+    protected function listsMemberships($id, array $params = [])
```
Comment (froschdesign) [inline library/ZendService/Twitter/Twitter.php]: [Duplicate](https://github.com/zendframework/ZendService_Twitter/blob/master/library/ZendService/Twitter/Twitter.php#L1294)

File: library/ZendService/Twitter/Twitter.php [hunk 2]
```diff
@@ -1355,4 +1355,65 @@ protected function friendsIds($id, array $params = [])
...
+     * Returns the relationships of the authenticating user to the 
+     * comma-separated list of up to 100 screen_names or user_ids provided.
+     *
+     * Returns the next cursor if there are more to be returned.
+     *
+     * @param  int|string $id
+     * @param  array $params
+     * @return Response
+     */
+    protected function friendshipsLookup($id, array $params = [])
```
Comment (froschdesign) [inline library/ZendService/Twitter/Twitter.php]: [Duplicate](https://github.com/zendframework/ZendService_Twitter/blob/master/library/ZendService/Twitter/Twitter.php#L1312)

**Comment by froschdesign** [comment]
The new features are documented here:

- [GET lists/subscribers](https://dev.twitter.com/rest/reference/get/lists/subscribers)
- [API Rate Limits](https://dev.twitter.com/rest/public/rate-limiting)

**Comment by calevans** [comment]
As good as it's gonna get. :)

File: library/ZendService/Twitter/Response.php [hunk 3]
```diff
@@ -51,21 +57,23 @@ class Response
...
-                'Unable to decode response from twitter: %s',
-                $e->getMessage()
-            ), 0, $e);
+
+        if (! is_null($httpResponse) and ! empty($httpResponse->getBody())) {
+            $this->populate($httpResponse);
         }
     }
 
+    protected function setRateLimit($headers = [])
```
Comment (Ocramius) [inline library/ZendService/Twitter/Response.php]: `private`. Also add docblock and type-hint to `$headers`

File: library/ZendService/Twitter/Response.php [hunk 4]
```diff
@@ -43,6 +43,12 @@ class Response
     protected $rawBody;
 
     /**
+     * @var array
+     */
+    protected $rateLimit;
```
Comment (Ocramius) [inline library/ZendService/Twitter/Response.php]: Docblock states that this is `array`, but according to constructor logic, this might be also `null`

File: library/ZendService/Twitter/Response.php [hunk 5]
```diff
@@ -51,21 +57,23 @@ class Response
...
-        try {
-            $jsonBody = Json::decode($this->rawBody, Json::TYPE_OBJECT);
-            $this->jsonBody = $jsonBody;
-        } catch (JsonException $e) {
-            throw new Exception\DomainException(sprintf(
-                'Unable to decode response from twitter: %s',
-                $e->getMessage()
-            ), 0, $e);
+
+        if (! is_null($httpResponse) and ! empty($httpResponse->getBody())) {
```
Comment (Ocramius) [inline library/ZendService/Twitter/Response.php]: `null === $httpResponse` or `! $httpResponse`. If you are checking for an empty response body, please also use `'' === $httpResponse->getBody()` instead of `empty()` (which would match also `0`, `'0'`, `false`, `[]` and `null`)

File: library/ZendService/Twitter/Response.php [hunk 6]
```diff
@@ -51,21 +57,23 @@ class Response
...
+            $this->populate($httpResponse);
         }
     }
 
+    protected function setRateLimit($headers = [])
+    {
+        $this->rateLimit['limit'] = isset($headers['x-rate-limit-limit']) ? $headers['x-rate-limit-limit'] : 0;
```
Comment (Ocramius) [inline library/ZendService/Twitter/Response.php]: Please use a single assignment here: `$this->rateLimit = [ 'key1' => expression, 'key2' => expression]`. Adding keys programmatically makes the code harder to understand, as it could lead to the assumption that different keys are set

File: library/ZendService/Twitter/Response.php [hunk 7]
```diff
@@ -51,21 +57,23 @@ class Response
...
+        $this->rateLimit['remaining'] = isset($headers['x-rate-limit-remaining']) ? $headers['x-rate-limit-remaining'] : 0;
+        $this->rateLimit['reset'] = isset($headers['x-rate-limit-reset']) ? $headers['x-rate-limit-reset'] : 0;
+        return;
```
Comment (Ocramius) [inline library/ZendService/Twitter/Response.php]: redundant

File: library/ZendService/Twitter/Response.php (hunk 4 above)
Comment (Ocramius) [inline library/ZendService/Twitter/Response.php]: `$rateLimit` seems to have a specific structure. Consider converting it into a tiny value object rather than using arrays

File: library/ZendService/Twitter/Response.php [hunk 8]
```diff
@@ -152,4 +160,52 @@ public function toValue()
     {
         return $this->jsonBody;
     }
+
+    /**
+     * Retun the array of response rate info
+     *
+     * @return array
+     */
+    public function getRateLimit()
```
Comment (Ocramius) [inline library/ZendService/Twitter/Response.php]: This method is usually not required. If it is required, then you should really not return an array, but rather a value object that represents the data-structure for rate limiting

File: library/ZendService/Twitter/Response.php [hunk 9]
```diff
@@ -152,4 +160,52 @@ public function toValue()
...
+    {
+        return $this->rateLimit;
+    }
+
+    /**
+     * Populates the object with info. This can possibly called from the
```
Comment (Ocramius) [inline library/ZendService/Twitter/Response.php]: Please don't make this method public: it is trivial to just call `__construct($httpResponse)`, so please don't allow the user to create an empty `Response` object, which may result in bugs if he forgets to call this method afterwards

File: library/ZendService/Twitter/Response.php [hunk 10]
```diff
@@ -152,4 +160,52 @@ public function toValue()
...
+            $jsonBody = Json::decode($this->rawBody, Json::TYPE_OBJECT);
+            $this->jsonBody = $jsonBody;
+        } catch (JsonException $e) {
+            throw new Exception\DomainException(sprintf(
+                'Unable to decode response from twitter: %s',
+                $e->getMessage()
+            ), 0, $e);
+        }
+
+        return;
```
Comment (Ocramius) [inline library/ZendService/Twitter/Response.php]: Redundant

File: library/ZendService/Twitter/Response.php [hunk 11]
```diff
@@ -152,4 +160,52 @@ public function toValue()
...
+            $httpResponse = $this->httpResponse;
+        }
+
+        $this->httpResponse = $httpResponse;
+        $this->rawBody      = $httpResponse->getBody();
+
+        $headers = $this->httpResponse->getHeaders();
+
+        if (! is_null($headers)) {
+            $headers = $headers->toArray();
```
Comment (Ocramius) [inline library/ZendService/Twitter/Response.php]: The semantic meaning of `$headers` seems to change during the method execution, as you are getting some sort of parameter bag and later on possibly converting it to an array. Consider defining two variables instead: that would also allow you to drop the `else` statement

File: library/ZendService/Twitter/Response.php [hunk 12]
```diff
@@ -152,4 +160,52 @@ public function toValue()
...
+        if (is_null($httpResponse)) {
+            $httpResponse = $this->httpResponse;
+        }
+
+        $this->httpResponse = $httpResponse;
+        $this->rawBody      = $httpResponse->getBody();
+
+        $headers = $this->httpResponse->getHeaders();
+
+        if (! is_null($headers)) {
```
Comment (Ocramius) [inline library/ZendService/Twitter/Response.php]: `null !== $headers`

File: library/ZendService/Twitter/Response.php [hunk 13]
```diff
@@ -152,4 +160,52 @@ public function toValue()
...
+
+    /**
+     * Populates the object with info. This can possibly called from the
+     * constructor, or it can be called later.
+     *
+     */
+    public function populate($httpResponse = null)
+    {
+
+        if (is_null($httpResponse)) {
```
Comment (Ocramius) [inline library/ZendService/Twitter/Response.php]: `! $httpResponse`. Also, I would disallow replacing the wrapped http response at runtime, since this object already represents a twitter-specific response value, and values don't really change once they are assigned

File: library/ZendService/Twitter/Twitter.php [hunk 14]
```diff
@@ -75,6 +75,13 @@ class Twitter
     protected $oauthConsumer = null;
 
     /**
+     * Response
+     *
+     * @var Twitter\Response
+     */
+    protected $twitterResponse = null;
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: `private`, unless there are strong reasons for making this protected. The `= null` is redundant`

File: library/ZendService/Twitter/Twitter.php [hunk 15]
```diff
@@ -75,6 +75,13 @@ class Twitter
     protected $oauthConsumer = null;
 
     /**
+     * Response
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: Doesn't express what this value is about

File: library/ZendService/Twitter/Twitter.php [hunk 16]
```diff
@@ -75,6 +75,13 @@ class Twitter
...
+     *
+     * @var Twitter\Response
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: `Twitter\Response|null`

File: library/ZendService/Twitter/Twitter.php [hunk 17]
```diff
@@ -115,8 +122,19 @@ class Twitter
...
-    public function __construct($options = null, OAuth\Consumer $consumer = null, Http\Client $httpClient = null)
-    {
+    public function __construct(
+        $options = null,
+        OAuth\Consumer $consumer = null,
+        Http\Client $httpClient = null,
+        Response $response = null
+    ) {
+        if (is_null($response)) {
+            $this->twitterResponse = new Response($httpClient);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: A response accepts an http response as constructor parameter, not an http client

File: library/ZendService/Twitter/Twitter.php [hunk 18]
```diff
@@ -318,7 +336,8 @@ public function accountVerifyCredentials()
     {
         $this->init();
         $response = $this->get('account/verify_credentials');
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: This goes hand in hand with the feedback above: please don't change the value at runtime, and just create a new `Response` object by just calling the constructor instead. Don't overwrite existing values by changing that internal value, as that just propagates shared mutable state, increasing bug likeliness.

File: library/ZendService/Twitter/Twitter.php [hunk 19]
```diff
@@ -333,7 +352,8 @@ public function applicationRateLimitStatus()
     {
         $this->init();
         $response = $this->get('application/rate_limit_status');
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 20]
```diff
@@ -350,7 +370,8 @@ public function blocksCreate($id)
         $path     = 'blocks/create';
         $params   = $this->createUserParameter($id, []);
         $response = $this->post($path, $params);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 21]
```diff
@@ -366,7 +387,8 @@ public function blocksDestroy($id)
         $path   = 'blocks/destroy';
         $params = $this->createUserParameter($id, []);
         $response = $this->post($path, $params);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 22]
```diff
@@ -382,7 +404,8 @@ public function blocksIds($cursor = -1)
         $this->init();
         $path = 'blocks/ids';
         $response = $this->get($path, ['cursor' => $cursor]);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 23]
```diff
@@ -398,7 +421,8 @@ public function blocksList($cursor = -1)
         $this->init();
         $path = 'blocks/list';
         $response = $this->get($path, ['cursor' => $cursor]);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 24]
```diff
@@ -415,7 +439,8 @@ public function directMessagesDestroy($id)
         $path     = 'direct_messages/destroy';
         $params   = ['id' => $this->validInteger($id)];
         $response = $this->post($path, $params);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 25]
```diff
@@ -460,7 +485,8 @@ public function directMessagesMessages(array $options = [])
             }
         }
         $response = $this->get($path, $params);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 26]
```diff
@@ -493,7 +519,8 @@ public function directMessagesNew($user, $text)
         $params         = $this->createUserParameter($user, []);
         $params['text'] = $text;
         $response       = $this->post($path, $params);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 27]
```diff
@@ -538,7 +565,8 @@ public function directMessagesSent(array $options = [])
             }
         }
         $response = $this->get($path, $params);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 28]
```diff
@@ -555,7 +583,8 @@ public function favoritesCreate($id)
         $path     = 'favorites/create';
         $params   = ['id' => $this->validInteger($id)];
         $response = $this->post($path, $params);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 29]
```diff
@@ -572,7 +601,8 @@ public function favoritesDestroy($id)
         $path     = 'favorites/destroy';
         $params   = ['id' => $this->validInteger($id)];
         $response = $this->post($path, $params);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 30]
```diff
@@ -621,7 +651,8 @@ public function favoritesList(array $options = [])
             }
         }
         $response = $this->get($path, $params);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 31]
```diff
@@ -645,7 +676,8 @@ public function friendshipsCreate($id, array $params = [])
         ];
         $params = array_intersect_key($params, $allowed);
         $response = $this->post($path, $params);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 32]
```diff
@@ -662,7 +694,8 @@ public function friendshipsDestroy($id)
         $path     = 'friendships/destroy';
         $params   = $this->createUserParameter($id, []);
         $response = $this->post($path, $params);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 33]
```diff
@@ -773,7 +806,8 @@ public function searchTweets($query, array $options = [])
             }
         }
         $response = $this->get($path, $params);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 34]
```diff
@@ -789,7 +823,8 @@ public function statusesDestroy($id)
         $this->init();
         $path = 'statuses/destroy/' . $this->validInteger($id);
         $response = $this->post($path);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 35]
```diff
@@ -847,7 +882,8 @@ public function statusesHomeTimeline(array $options = [])
             }
         }
         $response = $this->get($path, $params);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 36]
```diff
@@ -901,7 +937,8 @@ public function statusesMentionsTimeline(array $options = [])
...
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 37]
```diff
@@ -916,7 +953,8 @@ public function statusesSample()
         $this->init();
         $path = 'statuses/sample';
         $response = $this->get($path);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 38]
```diff
@@ -932,7 +970,8 @@ public function statusesShow($id)
         $this->init();
         $path = 'statuses/show/' . $this->validInteger($id);
         $response = $this->get($path);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 39]
```diff
@@ -970,7 +1009,8 @@ public function statusesUpdate($status, $inReplyToStatusId = null)
             $params['in_reply_to_status_id'] = $inReplyToStatusId;
         }
         $response = $this->post($path, $params);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 40]
```diff
@@ -1035,7 +1075,8 @@ public function statusesUserTimeline(array $options = [])
             }
         }
         $response = $this->get($path, $params);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 41]
```diff
@@ -1087,7 +1128,8 @@ public function usersSearch($query, array $options = [])
...
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 42]
```diff
@@ -1105,7 +1147,8 @@ public function usersShow($id)
         $path     = 'users/show';
         $params   = $this->createUserParameter($id, []);
         $response = $this->get($path, $params);
-        return new Response($response);
+        $this->twitterResponse->populate($response);
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: same as above: revert

File: library/ZendService/Twitter/Twitter.php [hunk 43]
```diff
@@ -1355,4 +1399,24 @@ protected function friendsIds($id, array $params = [])
...
+     * Returns the subscribers of the specified list. Private list subscribers
+     * will only be shown if the authenticated user owns the specified list.
+     *
+     * Returns the next cursor if there are more to be returned.
+     *
+     * @param  int|string $id
+     * @param  array $params
+     * @return Response
+     */
+    protected function listsSubscribers($id, array $params = [])
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: `private`, unless there are strong reasons to make it `protected`. If there are, then that needs to be documented in the docblock too

File: library/ZendService/Twitter/Twitter.php [hunk 44]
```diff
@@ -1355,4 +1399,24 @@ protected function friendsIds($id, array $params = [])
...
+    {
+        $this->init();
+        $path   = 'lists/subscribers';
```
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: Variable is redundant

File: tests/phpunit.xml.dist [hunk 45]
```diff
@@ -1,13 +0,0 @@
-<phpunit bootstrap="./Bootstrap.php" colors="true">
```
Comment (Ocramius) [inline tests/phpunit.xml.dist]: Possible rebase conflict? Or is this change intentional? See https://github.com/zendframework/ZendService_Twitter/commit/5320b6c1e8ba950556fd76144a694dd102049f1a

**Comment by calevans** [comment]
This is one of those patches that "works for me". I'm using my personal repo in production and I've submitted my changes back to the main repo. I'm fine if you don't want them but this code is in production and working fine.

The whole Response thing is simply so that it can be tested. I discussed it with MWOP and he was ok with it. That having been said, again, I'm ok if you don't want to commit it. I'm just sharing. :) 

Cheers! :)
=C=

**Comment by Ocramius** [comment]
> This is one of those patches that "works for me".

That argument is fine for your own fork, not for a patch sent to upstream.

I'm also fine with it if @weierophinney decides to abandon this package and move it under your own namespace, removing the sources from here and pointing links to your fork.

Until the repo lives under this organization, we're aiming at exceptional code quality and stability wherever possible, at least under the `Zend*` namespace, so the comments provided should either be debated and negated by discussion or fixed by new commits before a merge is considered.

File: library/ZendService/Twitter/Response.php (hunk 3 above)
Comment (weierophinney) [inline library/ZendService/Twitter/Response.php]: All other delegate methods in the class are protected, so this is actually consistent.

File: library/ZendService/Twitter/Response.php (hunk 4 above)
Comment (calevans) [inline library/ZendService/Twitter/Response.php]: Hi! :) 

To what end? Yes, it has a specific structure but adding another class to the mix won't really do anything. I would rather simply document the structure. 

=C=

File: library/ZendService/Twitter/Response.php (hunk 5 above)
Comment (calevans) [inline library/ZendService/Twitter/Response.php]: Hi! :)

Wouldn't any one of those be an error condition?

Cheers!
=C=

File: library/ZendService/Twitter/Response.php (hunk 6 above)
Comment (calevans) [inline library/ZendService/Twitter/Response.php]: Hi!

This seems to be largely a stylistic change. I prefer my style as I find 3 lines easier to read over the 1. However, if there is a pre-defined style guide that I am not aware of and have violated, I will change.

Cheers! 
=C=

File: library/ZendService/Twitter/Response.php (hunk 8 above)
Comment (calevans) [inline library/ZendService/Twitter/Response.php]: Hi! :)

Why is it better than an array?

The method is needed as the objects using the response need to be able to have access to this data. 

Cheers!
=C=

File: library/ZendService/Twitter/Response.php (hunk 9 above)
Comment (calevans) [inline library/ZendService/Twitter/Response.php]: Hi! :)

I am confused by the comment. getRateLimit has nothing to do with the __construct or the httpResponse. It's sole purpose it to allow read-only access tot he protected array rateLimit.  

Can you please clarify?

Cheers!
=C=

File: library/ZendService/Twitter/Response.php (hunk 11 above)
Comment (calevans) [inline library/ZendService/Twitter/Response.php]: Hi! :)

I can see where that may be true. I can change this to set a holding variable and then the final ->toArray() is stored in $headers. Personally, I find that to be more confusing as you are introducing an extra variable. However, this is a personal style preference. I don't find it any clearer to read and it would execute almost exactly the same. 

Cheers!
=C=

**Comment by calevans** [comment]
Ok, I've moved RateLimit into an object. I also updated the unit test.

File: library/ZendService/Twitter/Twitter.php (hunk 18 above)
Comment (calevans) [inline library/ZendService/Twitter/Twitter.php]: Hi! :)

This change was put into place - in all the place - so that it could be unit tested. Now instead of creating a Response object, we are injecting one and then populating it. This makes it possible to test.

This applied to all the ones below too. 

Cheers!
=C=

File: library/ZendService/Twitter/Response.php (hunk 5 above)
Comment (Ocramius) [inline library/ZendService/Twitter/Response.php]: Yes, any of those would be an error condition, but things like `'0'` would still pass json decoding.

File: library/ZendService/Twitter/Response.php (hunk 9 above)
Comment (Ocramius) [inline library/ZendService/Twitter/Response.php]: this comment is about `populate()`, which should only happen at `__construct` (or lazily, later), but should not be exposed to userland

File: library/ZendService/Twitter/Twitter.php (hunk 18 above)
Comment (Ocramius) [inline library/ZendService/Twitter/Twitter.php]: A response is a value object, that doesn't need replacement via a mock.

In this scenario, a proper test would do something like following (more assertions on the request needed):

```php
public function testAccountVerifyCredentials()
{
    $response = new HttpResponse();
    $response->setBody('the actual json here');
    $client = $this->createMock(HttpClient::class);
    $client->expects(self::once())->method('send')->with(self::isInstanceOf(Request::class))->willReturn($response);
    self::assertEquals(new Response($response), $this->twitter->accountVerifyCredentials());
}
```
